src/carbon_calculator/CCDefaults.py

```diff
@@ -25,56 +22,63 @@ def getLocality(inputs):
-                f = False
-                for i in range(len(DefaultsByLocality[c.locality][c.variable]["values"])):
-                    valid_date = DefaultsByLocality[c.locality][c.variable]["valid_dates"][i]
-                    if date < valid_date:
-                        # insert value at this point
-                        f = True
-                        DefaultsByLocality[c.locality][c.variable]["valid_dates"].insert(i,date)
-                        DefaultsByLocality[c.locality][c.variable]["values"].insert(i,c.value)
-                        break
-                    elif date == valid_date:
-                        # multiple values with one date
-                        print('CCDefaults: multiple values with same valid date')
-                        f = True
-                        break
-                
-                # if not inserted into list, append to the end
-                if not f:
-                    DefaultsByLocality[c.locality][c.variable]["valid_dates"].append(date)
-                    DefaultsByLocality[c.locality][c.variable]["values"].append(c.value)
-
-
-    except Exception as e:
-        print(str(e))
-        print("CalcDefault initialization skipped")
 
-    def __init__(self):
-        print("CCD __init__ called")
+    def ready(self):
+        self.DefaultsByLocality = {"default":{}} # the class variable
+        try:
+            cq = CalcDefault.objects.all()
+            for c in cq:
+                # valid date is 0 if not specified
+                date = '2000-01-01'
+                if c.valid_date != None:
+                    date = c.valid_date
+
+                if c.locality not in self.DefaultsByLocality:
+                    self.DefaultsByLocality[c.locality] = {}
+                if c.variable not in self.DefaultsByLocality[c.locality]:
+                    self.DefaultsByLocality[c.locality][c.variable] = {"valid_dates":[date], "values":[c.value]}
+                else:
+                    # already one value for this parameter, order by dates
+                    f = False
+                    for i in range(len(self.DefaultsByLocality[c.locality][c.variable]["values"])):
+                        valid_date = self.DefaultsByLocality[c.locality][c.variable]["valid_dates"][i]
+                        if date < valid_date:
+                            # insert value at this point
+                            f = True
+                            self.DefaultsByLocality[c.locality][c.variable]["valid_dates"].insert(i,date)
+                            self.DefaultsByLocality[c.locality][c.variable]["values"].insert(i,c.value)
+                            break
+                        elif date == valid_date:
+                            # multiple values with one date; clean this from database
+                            print("Duplicate encountered")
+                            f = True
+                            break
```

Comment:
this function is long and cumbersome with no clear explantion of what is happening.  
It will be extremely difficult to test and quite tidious to follow and give a proper review.  

I would appreciate it if we split it into smaller helper functions with good names (and comments) that get called in this main one.

We will be able to test the different aspects of it better that way.

Reply:
cleaned.  comments attempt to explain it.
the routine has been exhaustively tested and used for years, don't have time to rewrite it now.